### PR TITLE
feat: support typeref in RecordUtils.

### PR DESCRIPTION
**Pull request by RealChrisL**
## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: dao-api/src/test/java/com/linkedin/metadata/dao/utils/RecordUtilsTest.java [hunk 1]
```diff
@@ -149,6 +150,17 @@ public void testGetSelectedRecordTemplateFromUnion() throws IOException {
...
 
+  @Test
+  public void testGetSelectedRecordTemplateFromUnionWithTypeRef() throws IOException {
+    AspectBaz baz = new AspectBaz();
+    baz.setUnionField(new AspectBaz.UnionField());
+    baz.getUnionField().setTyperefPizzaInfo(new PizzaInfo().setMadeBy("bar"));
+
+    RecordTemplate selected = RecordUtils.getSelectedRecordTemplateFromUnion(baz.getUnionField());
+
+    assertEquals(selected.getClass(), PizzaInfo.class);
```
Comment (keremsahin1) [inline dao-api/src/test/java/com/linkedin/metadata/dao/utils/RecordUtilsTest.java:161]: Maybe also check the value of the record itself?

File: dao-api/src/main/java/com/linkedin/metadata/dao/utils/RecordUtils.java [hunk 2]
```diff
@@ -320,12 +321,15 @@ public static RecordTemplate toRecordTemplate(@Nonnull String className, @Nonnul
   public static <V extends RecordTemplate> RecordTemplate getSelectedRecordTemplateFromUnion(
       @Nonnull UnionTemplate unionTemplate) {
 
-    final DataSchema dataSchema = unionTemplate.memberType();
-    if (!(dataSchema instanceof RecordDataSchema)) {
+    DataSchema dataSchema = unionTemplate.memberType();
+    if (!(dataSchema instanceof RecordDataSchema) && !(dataSchema instanceof TyperefDataSchema)) {
       throw new InvalidSchemaException(
           "The currently selected member isn't a RecordTemplate in " + unionTemplate.getClass().getCanonicalName());
```
Comment (keremsahin1) [inline dao-api/src/main/java/com/linkedin/metadata/dao/utils/RecordUtils.java:327]: Fix exception message: RecordTemplate or TypeRef?

Comment (RealChrisL) [reply dao-api/src/main/java/com/linkedin/metadata/dao/utils/RecordUtils.java]: Nice catch, update the message.

File: dao-api/src/test/java/com/linkedin/metadata/dao/utils/RecordUtilsTest.java (hunk 1 above)
Comment (RealChrisL) [reply dao-api/src/test/java/com/linkedin/metadata/dao/utils/RecordUtilsTest.java:161]: Sounds good, use reflection to get the value.

File: dao-api/src/test/java/com/linkedin/metadata/dao/utils/RecordUtilsTest.java [hunk 3]
```diff
@@ -149,6 +150,18 @@ public void testGetSelectedRecordTemplateFromUnion() throws IOException {
...
+  @Test
+  public void testGetSelectedRecordTemplateFromUnionWithTypeRef() throws Exception {
+    AspectBaz baz = new AspectBaz();
+    baz.setUnionField(new AspectBaz.UnionField());
+    baz.getUnionField().setTyperefPizzaInfo(new PizzaInfo().setMadeBy("bar"));
+
+    RecordTemplate selected = RecordUtils.getSelectedRecordTemplateFromUnion(baz.getUnionField());
+
+    assertEquals(selected.getClass(), PizzaInfo.class);
+    assertEquals(selected.getClass().getMethod("getMadeBy").invoke(selected), "bar");
```
Comment (keremsahin1) [inline dao-api/src/test/java/com/linkedin/metadata/dao/utils/RecordUtilsTest.java:162]: What is wrong with `assertEquals(selected, new PizzaInfo().setMadeBy("bar"))` ?

Comment (RealChrisL) [reply dao-api/src/test/java/com/linkedin/metadata/dao/utils/RecordUtilsTest.java:162]: This is also a valid assertion. One is comparing the value in the `madeBy`, the other is comparing the PizzaInfo object.
Added the assertion in the test.

**Review by keremsahin1 — APPROVED** [review]

**Pull request merged**